applications/ShapeOptimizationApplication/python_scripts/algorithm_penalized_projection.py

```diff
@@ -94,96 +114,85 @@ def RunOptimizationLoop( self ):
         self.Analyzer.FinalizeAfterOptimizationLoop()
 
     # --------------------------------------------------------------------------
-    def __initializeNewShape( self ):
-        self.ModelPartController.UpdateMeshAccordingInputVariable( SHAPE_UPDATE )
+    def __initializeNewShape(self):
+        self.ModelPartController.UpdateMeshAccordingInputVariable(SHAPE_UPDATE)
         self.ModelPartController.SetReferenceMeshToMesh()
 
     # --------------------------------------------------------------------------
-    def __analyzeShape( self ):
+    def __analyzeShape(self):
         self.Communicator.initializeCommunication()
-        self.Communicator.requestValueOf( self.onlyObjectiveId )
-        self.Communicator.requestValueOf( self.onlyConstraintId )
-        self.Communicator.requestGradientOf( self.onlyObjectiveId )
-        self.Communicator.requestGradientOf( self.onlyConstraintId )
+        self.Communicator.requestValueOf(self.only_obj["identifier"].GetString())
+        self.Communicator.requestGradientOf(self.only_obj["identifier"].GetString())
+        self.Communicator.requestValueOf(self.only_con["identifier"].GetString())
+        self.Communicator.requestGradientOf(self.only_con["identifier"].GetString())
 
-        self.Analyzer.AnalyzeDesignAndReportToCommunicator( self.DesignSurface, self.optimizationIteration, self.Communicator )
+        self.Analyzer.AnalyzeDesignAndReportToCommunicator(self.DesignSurface, self.optimizationIteration, self.Communicator)
 
-        self.__storeResultOfSensitivityAnalysisOnNodes()
-        self.__RevertPossibleShapeModificationsDuringAnalysis()
+        objGradientDict = self.Communicator.getStandardizedGradient(self.only_obj["identifier"].GetString())
+        conGradientDict = self.Communicator.getStandardizedGradient(self.only_con["identifier"].GetString())
 
-    # --------------------------------------------------------------------------
-    def __storeResultOfSensitivityAnalysisOnNodes( self ):
-        gradientOfObjectiveFunction = self.Communicator.getStandardizedGradient( self.onlyObjectiveId )
-        gradientOfConstraintFunction = self.Communicator.getStandardizedGradient( self.onlyConstraintId )
-        self.__storeGradientOnNodalVariable( gradientOfObjectiveFunction, OBJECTIVE_SENSITIVITY )
-        self.__storeGradientOnNodalVariable( gradientOfConstraintFunction, CONSTRAINT_SENSITIVITY )
+        WriteDictionaryDataOnNodalVariable(objGradientDict, self.OptimizationModelPart, DF1DX)
+        WriteDictionaryDataOnNodalVariable(conGradientDict, self.OptimizationModelPart, DC1DX)
 
-    # --------------------------------------------------------------------------
-    def __storeGradientOnNodalVariable( self, gradient, variable_name ):
-        for nodeId, tmp_gradient in gradient.items():
-            self.OptimizationModelPart.Nodes[nodeId].SetSolutionStepValue(variable_name,0,tmp_gradient)
+        if self.only_obj["project_gradient_on_surface_normals"].GetBool():
```

Comment:
:+1: for allowing this for each response independently.
Is there a default_parameters for the objective / constraints tha can be used to validate the input?

Reply:
Specifying a default parameter here would be very hacky in the current implementation as the corresponding group in the settings ("objectives" or "constraints") is in many different places in the code. So the question is, where to check and validate the default settings for it. But indeed this is an important point and I added it to the project page.